.adr/0016-closing-go-routines.md

```diff
@@ -0,0 +1,134 @@
+
+A common and simple pattern we often use is simply closing the chan the go-routine is consuming from.
+
+```golang
+	toRoutine := make(chan int)
+
+	go func() {
+		for v := range toRoutine {
+			doSomething(v)
+		}
+		// This logic runs once the channel is closed
+		doCleanup()
+	}()
+
+	// This closes the channel and signals the go routine to exit
+	close(toRoutine)
+
+```
+
+This works, however a [cancelable context](https://cs.opensource.google/go/go/+/go1.20.5:src/context/context.go;l=238) provides some benefits over this:
+
+- A buffered chan will only get closed once it's buffer is read. This means a go-routine will read all the buffered entries before it finishes executing. By using a context we can halt the execution almost immediately.
+- It makes go-routine cleanup logic explicit and easy to see. It's now just a case statement for `ctx.Done`.
+- Minor but it would allow us to use other context features, such as timeouts, in the future.
+
+Due to these benefits, and the limited use of go-routines, we should update our structs to use a cancelable context to signal to a go-routine to stop executing.
+
+```golang
+	toRoutine := make(chan int)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case v := <-toRoutine:
+				doSomething(v)
+			case <-ctx.Done():
+				doCleanup()
+			}
+		}
+	}()
+
+	// This triggers the goroutine to exit
+	cancel()
+```
+
+### 2: Wait until all go-routines have completed execution.
+
+After we have signalled our "long-running" go-routines to exit we should wait for **all** go-routines to exit(both long and short lived). The easiest way to accomplish this is with a `sync.WaitGroup`
+
+```golang
+wg := sync.WaitGroup{}
+	toRoutine := make(chan int)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wg.Add(1)
+	go func() {
+		for {
+			select {
+			case v := <-toRoutine:
+				doSomething(v)
+			case <-ctx.Done():
+				doCleanup()
+				wg.Done()
+			}
+		}
+	}()
+
+	// This triggers the goroutine to exit
+	cancel()
+
+	// This blocks until the goroutine has exited
+	wg.Wait()
+```
+
+### Step 3: Close Resources
+
+Once a struct has waited for all go-routines to finish executing, it can dispose of any resources like network connections or child structs. We do this by calling `Close` on any child structs that implement [io.Closer interface](https://pkg.go.dev/io#Closer). **In general, if a child struct implements the `Closer` interface, we should consider calling it in our struct's `Close`**
+
+### Step 4: Return any errors
+
+If any error has
```

Comment:
This just trails off...?

Reply:
Oops, fixed in https://github.com/statechannels/go-nitro/pull/1343/commits/2a8e6e2dcd2c2b9c1e5cb3ea45fb09edd32ee04d